File: Algorithm/C++/131.palindrome-partitioning.cpp
/*
 * @lc app=leetcode id=131 lang=cpp
 *
 * [131] Palindrome Partitioning
 */

// @lc code=start

#include <vector>
#include <string>

class Solution {
public:
    std::vector<std::vector<std::string>> partition(std::string s) {
        result.clear();
        candidate.clear();
        backtrack(s, 0);
        return result;
    }

private:
    bool isPalindorme(std::string& s, int left, int right) {
        while(left < right) {
            if (s[left++] != s[right--]) return false;
        }
        return true;
    }

    void backtrack(std::string& s, int idx) {
        if (idx == s.size()) {
            result.push_back(candidate);
            return;
        }
        for (int i = idx; i < s.size(); i++) {
            if (isPalindorme(s, idx, i)) {
                candidate.push_back(s.substr(idx, i - idx + 1));
                backtrack(s, i+1);
                candidate.pop_back();
            }
        }
    }

    std::vector<std::vector<std::string>> result;
    std::vector<std::string> candidate;
};

// @lc code=end

File: Algorithm/Python/131.palindrome-partitioning.py
#
# @lc app=leetcode id=131 lang=python
#
# [131] Palindrome Partitioning
#

# @lc code=start
class Solution(object):
    def partition(self, s):
        """
        :type s: str
        :rtype: List[List[str]]
        """
        result, n = [], len(s)
        dp = [[False]*n for _ in range(n)]
        for i in range(n-1, -1, -1):
            for j in range(i, n, 1):
                if i == j:
                    dp[i][j] = True
                elif s[i] == s[j] and (i+1 >= j-1 or dp[i+1][j-1]):
                    dp[i][j] = True
        
        def dfs(idx, tmp):
            if idx == n:
                result.append(tmp[:])
                return
            for i in range(idx, n):
                if dp[idx][i]:
                    tmp.append(s[idx:i+1])
                    dfs(i+1, tmp)
                    tmp.pop()
        
        dfs(0, [])
        return result



# @lc code=end

